pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,474 @@
+	"github.com/gardener/gardener-extension-provider-openstack/pkg/openstack"
+	openstackclient "github.com/gardener/gardener-extension-provider-openstack/pkg/openstack/client"
+
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	ctrlerror "github.com/gardener/gardener/pkg/controllerutils/reconciler"
+	"github.com/go-logr/logr"
+	"github.com/gophercloud/gophercloud"
+	computefip "github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/floatingips"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/layer3/floatingips"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/security/groups"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/security/rules"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// bastionEndpoints holds the endpoints the bastion host provides
+type bastionEndpoints struct {
+	// private is the private endpoint of the bastion. It is required when opening a port on the worker node to allow SSH access from the bastion
+	private *corev1.LoadBalancerIngress
+	//  public is the public endpoint where the enduser connects to establish the SSH connection.
+	public *corev1.LoadBalancerIngress
+}
+
+// ready returns true if both public and private interfaces each have either
+// an IP or a hostname or both.
+func (be *bastionEndpoints) ready() bool {
+	return be != nil && IngressReady(be.private) && IngressReady(be.public)
+}
+
+func (a *actuator) Reconcile(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	logger := a.logger.WithValues("bastion", client.ObjectKeyFromObject(bastion), "operation", "reconcile")
+
+	err := bastionConfigCheck(a.bastionConfig)
+	if err != nil {
+		return err
+	}
+
+	opt, err := DetermineOptions(bastion, cluster)
+	if err != nil {
+		return err
+	}
+
+	credentials, err := openstack.GetCredentials(ctx, a.client, opt.SecretReference, false)
+	if err != nil {
+		return fmt.Errorf("could not get Openstack credentials: %w", err)
+	}
+
+	openstackClientFactory, err := a.openstackClientFactory.NewFactory(credentials)
+	if err != nil {
+		return fmt.Errorf("could not create Openstack client factory: %w", err)
+	}
+
+	computeClient, err := openstackClientFactory.Compute()
+	if err != nil {
+		return err
+	}
+
+	networkingClient, err := openstackClientFactory.Networking()
+	if err != nil {
+		return err
+	}
+
+	securityGroup, err := ensureSecurityGroup(networkingClient, opt)
+	if err != nil {
+		return err
+	}
+
+	err = ensureSecurityGroupRules(networkingClient, bastion, opt, securityGroup.ID)
+	if err != nil {
+		return err
+	}
+
+	networkInfo, err := networkingClient.GetNetworkByName(opt.ShootName)
```

Comment:
as with the other providers (e.g. https://github.com/gardener/gardener-extension-provider-gcp/pull/436) its not safe to assume that the network will always have the name of the shoot and you need to fetch the information from the infrastructure status.